storage/queue.go

```diff
@@ -155,21 +165,84 @@ func (bq *baseQueue) MaybeAdd(now proto.Timestamp, rng *Range) {
 	if pqLen := bq.priorityQ.Len(); pqLen > bq.maxSize {
 		bq.remove(pqLen - 1)
 	}
+	// Signal the processLoop that a range has been added.
+	bq.incoming <- rng
 }
 
 // MaybeRemove removes the specified range from the queue if enqueued.
 func (bq *baseQueue) MaybeRemove(rng *Range) {
+	bq.Lock()
+	defer bq.Unlock()
 	if item, ok := bq.ranges[rng.Desc.RaftID]; ok {
+		log.Infof("removing range %s from %s queue", item.value, bq.name)
 		bq.remove(item.index)
 	}
 }
 
-// Clear removes all ranges from the queue.
-func (bq *baseQueue) Clear() {
-	bq.ranges = map[int64]*rangeItem{}
-	bq.priorityQ = nil
+// process processes the entries in the queue until the provided
+// stopper signals exit.
+//
+// TODO(spencer): current load should factor into range processing timer.
+func (bq *baseQueue) processLoop(clock *hlc.Clock, stopper *util.Stopper) {
+	// nextTime is set arbitrarily far into the future so that we don't
+	// unecessarily check for a range to dequeue if the timer function
+	// returns a short duration but the priority queue is empty.
+	emptyQueue := true
+	nextTime := time.Now().Add(24 * time.Hour)
```

Comment:
I'm using time directly here for select {} timers. More than just encapsulating the time package would be required to make testing here depend on a logical clock. I'm open to suggestions though.